### ISSUE TITLE
Support configurable max nested level for wafv2 rule statement

### ISSUE THREAD

**Issue by MAXxATTAXx** [terraform-providers/terraform-provider-aws#14377]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Allow during provider definition on terraform code to set the max level of nested statement supported by the schema.
Currently that value is hard coded to 3 max levels which excludes more complex use cases of the statement definition.

### New or Affected Resource(s)

* aws_wafv2_rule_group
* aws_wafv2_web_acl

### Potential Terraform Configuration

```hcl
### Core rules
resource "aws_wafv2_rule_group" "core" {
  name        = "core"
  description = "Contains rules that are applicable for most web applications. (OWASP and more)"
  capacity    = 200
  scope       = "REGIONAL"

  ## SQLI

  ## XSS
  # body
  rule {
    name     = "xss_body"
    priority = 20

    action {
      block {}
    }

    statement {
      and_statement {
        # XSS inspection
        statement {
          xss_match_statement {
            field_to_match {
              body {}
            }

            text_transformation {
              priority = 0
              type     = "HTML_ENTITY_DECODE"
            }

            text_transformation {
              priority = 1
              type     = "URL_DECODE"
            }
          }
        }

        # Whitelist
         statement {
           not_statement {
             statement {
               and_statement {
                 statement {
                   byte_match_statement {
                     positional_constraint = "CONTAINS"
                     search_string         = "API-DOMAIN1"

                     field_to_match {
                       single_header {
                         name = "host"
                       }
                     }

                     text_transformation {
                       priority = 0
                       type     = "LOWERCASE"
                     }
                   }
                 }

                 statement {
                   byte_match_statement {
                     positional_constraint = "CONTAINS"
                     search_string         = "API-DOMAIN2"

                     field_to_match {
                       single_header {
                         name = "host"
                       }
                     }

                     text_transformation {
                       priority = 0
                       type     = "LOWERCASE"
                     }
                   }
                 }
               }
             }
           }
         }
      }
    }

    visibility_config {
      cloudwatch_metrics_enabled = true
      metric_name                = "xss_body"
      sampled_requests_enabled   = true
    }
  }

  visibility_config {
    cloudwatch_metrics_enabled = true
    metric_name                = "core"
    sampled_requests_enabled   = true
  }
}
```

**Comment by anGie44** [comment]
Hi @MAXxATTAXx, thank you for creating this issue! while the AWS Go SDK / APIs do allow for more than 3 levels of nesting, unfortunately within the provider, we've come across #14062 in the web_acl resource by supporting essentially 1 root statement and 3 nested statements which creates a challenging practitioner experience. so at this time, development to support this feature is blocked but I'll update here if there are any changes that can enable this request in the near future.

**Comment by demolitionmode** [comment]
Hi @anGie44, https://github.com/hashicorp/terraform-provider-aws/issues/14062 seems to be resolved now, is there anything else that is blocking development on this?

**Comment by vbakaitis** [comment]
Hi @anGie44, just following up with a bit more details about this:

I hit the 3 level `statement` nesting limitation and tried to work around it by using `lifecycle` and adding `rule` to `ignore_changes`, however even that does not work, terraform produces this error when running `terraform plan` after adding more than 3 levels of nested statements via AWS console:

```
Error: Error setting rule: Invalid address to set: []string{"rule", "0", "statement", "0", "and_statement", "0", "statement", "1", "not_statement", "0", "statement", "0", "and_statement"}
```

At this point I'm planning to completely remove the offending rule group from terraform and manage it manually, but would love to hear about any workarounds or developments in this area.

**Comment by olahouze** [comment]
Hello, news from this issue ?

**Comment by breathingdust** [comment]
A workaround for this was added in [v5.61.0](https://github.com/hashicorp/terraform-provider-aws/blob/v5.61.0/CHANGELOG.md) that allows you to specify the raw json in order to use a larger nested rule. The upstream Terraform issue still persists, and does result in the resource being virtually unusable with rules with any more than 3 levels of nesting (literal hours). My understanding is that the upstream issue is not likely to be resolved in the short term. We will keep this open in case the upstream issue does get fixed.

**Comment by kcul** [comment]
Hello, any progress for this issue, or any plan on this. raw_json is not a practical workaround, since **no drift** for it, the terraform loss its meaning!

**Comment by staizen-stephen** [comment]
Just hit this and agree that this needs a solution. It's already a pretty difficult resource API. Hitting additional limitations is unfortunate 😞

**Comment by philomory** [comment]
I would even vote that the solution to this is that the entire block-based implementation should be deprecated, and `rule_json` be the official way of handling things going forward (including having drift detection), rather than a workaround. Besides the limitation around nesting noted in this ticket, the nested-block based API is just miserable to work with, especially when taking into account how clumsy and restrictive the upstream API can be; e.g. the fact that an `AndStatement` element _must_ contain _more than one_ `Statement` element, so if you want to supply "one or more" statements you end up with just a giant mess of dueling, nested `dynamic` blocks with common content that you must simply copy/paste because blocks can't be assigned to variables.

**Comment by ghost** [comment]
> I would even vote that the solution to this is that the entire block-based implementation should be deprecated, and `rule_json` be the official way of handling things going forward (including having drift detection), rather than a workaround. Besides the limitation around nesting noted in this ticket, the nested-block based API is just miserable to work with, especially when taking into account how clumsy and restrictive the upstream API can be; e.g. the fact that an `AndStatement` element _must_ contain _more than one_ `Statement` element, so if you want to supply "one or more" statements you end up with just a giant mess of dueling, nested `dynamic` blocks with common content that you must simply copy/paste because blocks can't be assigned to variables.

Absolutely. I rewrote a large number of rules and logic into the JSON object format using HCL syntax and multiple local variables. It's still pretty horrible to work with but it is now much more manageable overall. I won't be going back to the non-JSON approach.
